fix: prevent publish panic during unsubscribe

A downstream project reported a panic from `github.com/leandro-lugaresi/hub@v1.1.1`:

```text
panic: send on closed channel

github.com/leandro-lugaresi/hub.(*nonBlockingSubscriber).Set(...)
    subscriber.go:39
github.com/leandro-lugaresi/hub.(*Hub).Publish(...)
    hub.go:32
```

Reference stack trace included `github.com/photoprism/photoprism/internal/event.(*Hook).Fire` publishing through hub while serving an HTTP request.

This can happen when `Publish` races with `Unsubscribe` or `Close`.

Current flow:

```text
goroutine A: Hub.Publish calls matcher.Lookup and gets subscriber
goroutine B: Hub.Unsubscribe removes subscriber and closes its channel
goroutine A: subscriber.Set sends to the now-closed channel
```

Because both `blockingSubscriber.Set` and `nonBlockingSubscriber.Set` send directly to the channel, either subscriber type can panic if its channel is closed after lookup but before delivery.

## Expected behavior

Publishing to a stale subscriber should not panic. If a subscription has already been closed, the message should be ignored.

## Scope

Fix both subscriber implementations:

- `nonBlockingSubscriber`
- `blockingSubscriber`

`Set` and `Close` should be safe when called concurrently.

## Reproduction test

```go
func TestPublishDoesNotPanicWhenNonBlockingSubscriberClosesAfterLookup(t *testing.T) {
	for i := 0; i < 10000; i++ {
		h := New()
		sub := h.NonBlockingSubscribe(1, "log.*")
		panicCh := make(chan interface{}, 1)
		done := make(chan struct{})

		go func() {
			defer close(done)
			defer func() {
				if r := recover(); r != nil {
					panicCh <- r
				}
			}()

			for j := 0; j < 100; j++ {
				h.Publish(Message{Name: "log.debug"})
			}
		}()

		h.Unsubscribe(sub)
		<-done

		select {
		case p := <-panicCh:
			require.Failf(t, "Publish panicked", "iteration %d panic: %v", i, p)
		default:
		}
	}
}
```

## Acceptance criteria

- Concurrent `Publish` and `Unsubscribe` do not panic.
- Fix covers blocking and non-blocking subscribers.
- Add regression coverage for the race.
- Existing tests continue to pass.

--- a/hub_test.go
+++ b/hub_test.go
@@ -150,6 +150,51 @@ func TestWith(t *testing.T) {
 	require.Falsef(t, ok, "Unsubscribe should close the internal channel, received fields %v", v)
 }
 
+func TestPublishDoesNotPanicWhenNonBlockingSubscriberClosesAfterLookup(t *testing.T) {
+	testPublishDoesNotPanicWhenSubscriberClosesAfterLookup(t, func(h *Hub) Subscription {
+		return h.NonBlockingSubscribe(1, "log.*")
+	})
+}
+
+func TestPublishDoesNotPanicWhenBlockingSubscriberClosesAfterLookup(t *testing.T) {
+	testPublishDoesNotPanicWhenSubscriberClosesAfterLookup(t, func(h *Hub) Subscription {
+		return h.Subscribe(100, "log.*")
+	})
+}
+
+func testPublishDoesNotPanicWhenSubscriberClosesAfterLookup(t *testing.T, subscribe func(*Hub) Subscription) {
+	t.Helper()
+
+	for i := 0; i < 10000; i++ {
+		h := New()
+		sub := subscribe(h)
+		panicCh := make(chan interface{}, 1)
+		done := make(chan struct{})
+
+		go func() {
+			defer close(done)
+			defer func() {
+				if r := recover(); r != nil {
+					panicCh <- r
+				}
+			}()
+
+			for j := 0; j < 100; j++ {
+				h.Publish(Message{Name: "log.debug"})
+			}
+		}()
+
+		h.Unsubscribe(sub)
+		<-done
+
+		select {
+		case p := <-panicCh:
+			require.Failf(t, "Publish panicked", "iteration %d panic: %v", i, p)
+		default:
+		}
+	}
+}
+
 func newMessageCounter(s Subscription) *messageCounter {
 	ms := &messageCounter{sub: s, c: 0}
 	go func(ms *messageCounter) {

--- a/matching_cstrie.go
+++ b/matching_cstrie.go
@@ -528,7 +528,7 @@ func cleanParent(i, parent, parentsParent *iNode, c *csTrieMatcher, word string)
 			}
 
 			if main.tNode != nil {
-				if !contract(parentsParent, parent, c, pMain) {
+				if parentsParent != nil && !contract(parentsParent, parent, c, pMain) {
 					cleanParent(parentsParent, parent, i, c, word)
 				}
 			}

--- a/subscriber.go
+++ b/subscriber.go
@@ -11,11 +11,15 @@ type (
 		ch        chan Message
 		alert     alertFunc
 		onceClose sync.Once
+		mu        sync.RWMutex
+		closed    bool
 	}
 	// blockingSubscriber uses an channel to receive events.
 	blockingSubscriber struct {
 		ch        chan Message
 		onceClose sync.Once
+		mu        sync.RWMutex
+		closed    bool
 	}
 )
 
@@ -35,6 +39,13 @@ func newNonBlockingSubscriber(cap int, alerter alertFunc) *nonBlockingSubscriber
 
 // Set inserts the given Event into the diode.
 func (s *nonBlockingSubscriber) Set(msg Message) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if s.closed {
+		return
+	}
+
 	select {
 	case s.ch <- msg:
 	default:
@@ -50,6 +61,10 @@ func (s *nonBlockingSubscriber) Ch() <-chan Message {
 // Close will close the internal channel and stop receiving messages.
 func (s *nonBlockingSubscriber) Close() {
 	s.onceClose.Do(func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		s.closed = true
 		close(s.ch)
 	})
 }
@@ -67,6 +82,13 @@ func newBlockingSubscriber(cap int) *blockingSubscriber {
 
 // Set will send the message using the channel.
 func (s *blockingSubscriber) Set(msg Message) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if s.closed {
+		return
+	}
+
 	s.ch <- msg
 }
 
@@ -78,6 +100,10 @@ func (s *blockingSubscriber) Ch() <-chan Message {
 // Close will close the internal channel and stop receiving messages.
 func (s *blockingSubscriber) Close() {
 	s.onceClose.Do(func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		s.closed = true
 		close(s.ch)
 	})
 }

